src/aqua/decider.aqua

```diff
@@ -231,48 +231,48 @@ func need_update_from_block(listener_id: string, net: string, to_block: string)
---                         -- Take the latest change to apply it.
---                         latest_deal_update = result.result[number_of_changes - 1]
---                         app_cid <- Json.stringify(latest_deal_update.info.app_cid)
---                         -- send to worker
---                         on joined_deal.worker_id:
---                             Spell joined_deal.spell_id
---                             Spell.set_string("worker_def_cid", app_cid)
--- 
---                         -- Update the deal
---                         new_deal_state = DealState(from_block = latest_deal_update.block_number, app_cid = latest_deal_update.info.app_cid)
---                         store_deal_state(spell_id, joined_deal.deal_id, new_deal_state)
---                     else:
---                         -- If we found no deals, we check if need to move the from_block forward
---                         -- We need to do it in case if no deals were done in range of 10000 blocks
---                         need_update <- need_update_from_block(listener_id, net, result.to_block)
---                         if need_update:
---                             new_deal_state2 = DealState(from_block = result.to_block, app_cid = deal_state.app_cid)
---                             store_deal_state(spell_id, joined_deal.deal_id, new_deal_state2)
+func poll_deal_changes(spell_id: string, listener_id: string, net: string):
+    FluenceAuroraConnector listener_id
+    Spell spell_id
+    -- Get a list of all known deals
+    list <- Spell.list_get_strings("joined_deals")
+    if list.success:
+        for str <- list.strings:
+            joined_deal <- JsonJoinedDeal.parse(str)
+            -- Get detailed state of the deal
+            deal_str, is_ok <- get_string(spell_id, joined_deal.deal_id)
+            if is_ok == false:
+                log(spell_id, ["found joined_deal, but not saved state", joined_deal.deal_id])
+            else:
+                deal_state <- JsonDealState.parse(deal_str!)
+                -- Find updates for the deals
+                deal_address <- Op.concat_strings("0x", joined_deal.deal_id)
+                result <- FluenceAuroraConnector.poll_deal_change(net, deal_address, deal_state.from_block)
+                if result.success == false:
+                    log(spell_id, ["can't retrieve deal changes", joined_deal.deal_id, deal_state.from_block])
+                else:
+                    -- Find how many changes are there
+                    number_of_changes = result.result.length
+                    if number_of_changes > 0:
+                        log(spell_id, ["sending the latest update to the worker", joined_deal.worker_id, joined_deal.spell_id, "for deal", joined_deal.deal_id])
+                        -- Take the latest change to apply it.
+                        latest_deal_update = result.result[number_of_changes - 1]
+                        app_cid <- Json.stringify(latest_deal_update.info.app_cid)
```

Comment:
Why do you stringify? It's saved in `set_string` anyway 🤔

Reply:
Because it's used as an argument, it needs to be a json object.